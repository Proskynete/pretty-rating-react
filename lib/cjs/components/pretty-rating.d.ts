/// <reference types="react" />
import { PrettyRatingInterface } from '../types';
declare const PrettyRating: ({ rating, icons, iconsNumber, setColors, }: PrettyRatingInterface) => JSX.Element[];
export default PrettyRating;
