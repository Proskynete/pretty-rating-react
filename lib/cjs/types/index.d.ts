import { IconDefinition } from '@fortawesome/fontawesome-svg-core';
export interface AllowIconsType {
    complete: string | IconDefinition;
    half: string | IconDefinition;
    empty: string | IconDefinition;
}
export interface BaseInterface {
    rating: number | string;
    icons: AllowIconsType;
}
export interface PrettyRatingInterface extends BaseInterface {
    setColors?: string[];
    iconsNumber?: number;
}
export interface CreateIconInterface extends BaseInterface {
    iconsNumber: number;
}
export interface CustomStylesInterface {
    currentIcon: string | IconDefinition;
    icons: AllowIconsType;
    colors: string[];
    type: string;
}
export interface CreateIconResponseInterface {
    name: string | IconDefinition;
    type: string;
}
