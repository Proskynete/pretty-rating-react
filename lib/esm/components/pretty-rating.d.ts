import React from 'react';
import { PrettyRatingProps } from '../types';
declare const MemoizedPrettyRating: React.MemoExoticComponent<({ value, icons, max, colors, disabled, onChange, }: PrettyRatingProps) => JSX.Element>;
export { MemoizedPrettyRating as PrettyRating };
