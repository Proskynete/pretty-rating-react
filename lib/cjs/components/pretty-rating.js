"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PrettyRating = void 0;
var react_1 = __importStar(require("react"));
var fontawesome_svg_core_1 = require("@fortawesome/fontawesome-svg-core");
var react_fontawesome_1 = require("@fortawesome/react-fontawesome");
var free_solid_svg_icons_1 = require("@fortawesome/free-solid-svg-icons");
var create_icon_helper_1 = require("../helpers/create-icon.helper");
var custom_styles_helper_1 = require("../helpers/custom-styles.helper");
var config_1 = require("../config");
fontawesome_svg_core_1.library.add(free_solid_svg_icons_1.fas);
var PrettyRating = function (_a) {
    var value = _a.value, _b = _a.icons, icons = _b === void 0 ? config_1.defaultValues.icons : _b, _c = _a.max, max = _c === void 0 ? config_1.defaultValues.max : _c, _d = _a.colors, colors = _d === void 0 ? config_1.defaultValues.colors : _d, _e = _a.disabled, disabled = _e === void 0 ? config_1.defaultValues.disabled : _e, onChange = _a.onChange;
    var _f = (0, react_1.useState)(), iconsToRenderState = _f[0], setIconsToRenderState = _f[1];
    (0, react_1.useEffect)(function () {
        var iconsToRender = (0, create_icon_helper_1.createIcon)({ value: value, icons: icons, max: max });
        setIconsToRenderState(iconsToRender);
    }, []);
    var handleOnMouseEnter = function (i) {
        var iconsToRender = (0, create_icon_helper_1.createIcon)({ value: i, icons: icons, max: max });
        setIconsToRenderState(iconsToRender);
    };
    var handleOnMouseLeave = function () {
        var iconsToRender = (0, create_icon_helper_1.createIcon)({ value: value, icons: icons, max: max });
        setIconsToRenderState(iconsToRender);
    };
    var handleOnChange = function (newValue) {
        onChange && onChange(newValue);
    };
    return (react_1.default.createElement(react_1.default.Fragment, null, iconsToRenderState === null || iconsToRenderState === void 0 ? void 0 : iconsToRenderState.map(function (icon, i) {
        return (react_1.default.createElement(react_fontawesome_1.FontAwesomeIcon, { key: i, icon: typeof icon.name === 'string' ? icon.name : icon.name, style: {
                cursor: !disabled ? 'pointer' : 'default',
                color: (0, custom_styles_helper_1.customStyles)({
                    current: icon.name,
                    icons: icons,
                    type: icon.type,
                    colors: colors,
                }),
            }, onMouseEnter: function () { return !disabled && handleOnMouseEnter(i + 1); }, onMouseLeave: function () { return !disabled && handleOnMouseLeave(); }, onClick: function () { return !disabled && handleOnChange(i + 1); } }));
    })));
};
var MemoizedPrettyRating = (0, react_1.memo)(PrettyRating);
exports.PrettyRating = MemoizedPrettyRating;
