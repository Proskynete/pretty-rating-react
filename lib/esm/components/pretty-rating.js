import React, { memo, useEffect, useState } from 'react';
import { library } from '@fortawesome/fontawesome-svg-core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { fas } from '@fortawesome/free-solid-svg-icons';
import { createIcon } from '../helpers/create-icon.helper';
import { customStyles } from '../helpers/custom-styles.helper';
import { defaultValues } from '../config';
library.add(fas);
var PrettyRating = function (_a) {
    var value = _a.value, _b = _a.icons, icons = _b === void 0 ? defaultValues.icons : _b, _c = _a.max, max = _c === void 0 ? defaultValues.max : _c, _d = _a.colors, colors = _d === void 0 ? defaultValues.colors : _d, _e = _a.disabled, disabled = _e === void 0 ? defaultValues.disabled : _e, onChange = _a.onChange;
    var _f = useState(), iconsToRenderState = _f[0], setIconsToRenderState = _f[1];
    useEffect(function () {
        var iconsToRender = createIcon({ value: value, icons: icons, max: max });
        setIconsToRenderState(iconsToRender);
    }, []);
    var handleOnMouseEnter = function (i) {
        var iconsToRender = createIcon({ value: i, icons: icons, max: max });
        setIconsToRenderState(iconsToRender);
    };
    var handleOnMouseLeave = function () {
        var iconsToRender = createIcon({ value: value, icons: icons, max: max });
        setIconsToRenderState(iconsToRender);
    };
    var handleOnChange = function (newValue) {
        onChange && onChange(newValue);
    };
    return (React.createElement(React.Fragment, null, iconsToRenderState === null || iconsToRenderState === void 0 ? void 0 : iconsToRenderState.map(function (icon, i) {
        return (React.createElement(FontAwesomeIcon, { key: i, icon: typeof icon.name === 'string' ? icon.name : icon.name, style: {
                cursor: !disabled ? 'pointer' : 'default',
                color: customStyles({
                    current: icon.name,
                    icons: icons,
                    type: icon.type,
                    colors: colors,
                }),
            }, onMouseEnter: function () { return !disabled && handleOnMouseEnter(i + 1); }, onMouseLeave: function () { return !disabled && handleOnMouseLeave(); }, onClick: function () { return !disabled && handleOnChange(i + 1); } }));
    })));
};
var MemoizedPrettyRating = memo(PrettyRating);
export { MemoizedPrettyRating as PrettyRating };
