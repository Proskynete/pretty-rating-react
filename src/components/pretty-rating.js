import React from 'react';
import PropTypes from 'prop-types';
import { createIcon } from '../helper/create-icon.helper';
import { customStyles } from '../helper/custom-styles.helper';

const PrettyRating = (props) => {
	const { rating, icons, iconsNumber, setColors } = props;
	const iconsToRender = createIcon({ rating, icons, iconsNumber });

	return iconsToRender.map((icon) => {
		return (
			<i
				key={Math.random()}
				style={{
					color: customStyles({
						currentIcon: icon.name,
						icons,
						type: icon.type,
						colors: setColors,
					}),
				}}
				className={icon.name}
				aria-hidden='true'
			/>
		);
	});
};

PrettyRating.defaultProps = {
	iconsNumber: 5,
	setColors: ['#000', '#000', '#000'],
};

PrettyRating.propTypes = {
	rating: PropTypes.oneOfType([PropTypes.number, PropTypes.string]).isRequired,
	icons: PropTypes.shape({ complete: '', half: '', empty: '' }).isRequired,
	setColors: PropTypes.instanceOf(Array),
	iconsNumber: PropTypes.number,
};

export default PrettyRating;
