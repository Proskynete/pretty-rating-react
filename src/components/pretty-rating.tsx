import React, { useEffect, useState } from 'react';
import { IconProp, library } from '@fortawesome/fontawesome-svg-core';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { fas } from '@fortawesome/free-solid-svg-icons';
import { createIcon } from '../helpers/create-icon.helper';
import { customStyles } from '../helpers/custom-styles.helper';
import * as PrettyRatingInterface from '../types';
import { defaultValues } from '../config';

library.add(fas);

export const PrettyRating = ({
	value,
	icons = defaultValues.icons,
	max = defaultValues.max,
	colors = defaultValues.colors,
	onChange,
}: PrettyRatingInterface.Props) => {
	const [iconsToRenderState, setIconsToRenderState] =
		useState<PrettyRatingInterface.CreateIconResponse[]>();

	useEffect(() => {
		const iconsToRender = createIcon({ value, icons, max });
		setIconsToRenderState(iconsToRender);
	}, []);

	const handleOnMouseEnter = (i: number) => {
		const iconsToRender = createIcon({ value: i, icons, max });
		setIconsToRenderState(iconsToRender);
	};

	const handleOnMouseLeave = () => {
		const iconsToRender = createIcon({ value, icons, max });
		setIconsToRenderState(iconsToRender);
	};

	const handleOnChange = (newValue: number) => {
		onChange && onChange(newValue);
	};

	return (
		<>
			{iconsToRenderState?.map((icon, i) => {
				return (
					<FontAwesomeIcon
						key={i}
						icon={typeof icon.name === 'string' ? (icon.name as IconProp) : icon.name}
						style={{
							cursor: 'pointer',
							color: customStyles({
								current: icon.name,
								icons,
								type: icon.type,
								colors,
							}),
						}}
						onMouseEnter={() => handleOnMouseEnter(i + 1)}
						onMouseLeave={() => handleOnMouseLeave()}
						onClick={() => handleOnChange(i + 1)}
					/>
				);
			})}
		</>
	);
};
