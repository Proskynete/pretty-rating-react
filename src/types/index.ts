export interface AllowIconsType {
	complete: string;
	half: string;
	empty: string;
}

export interface BaseInterface {
	rating: number | string;
	icons: AllowIconsType;
}

export interface PrettyRatingInterface extends BaseInterface {
	setColors?: string[];
	iconsNumber?: number;
}

export interface CreateIconInterface extends BaseInterface {
	iconsNumber: number;
}

export interface CustomStylesInterface {
	currentIcon: string;
	icons: AllowIconsType;
	colors: string[];
	type: string;
}

export interface CreateIconResponseInterface {
	name: string;
	type: string;
}
