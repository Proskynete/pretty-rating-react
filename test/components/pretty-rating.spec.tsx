import React from 'react';
import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import renderer from 'react-test-renderer';

import { faStar, faStarHalfAlt } from '@fortawesome/free-solid-svg-icons';
import { AllowIconsType } from '../../src/types';
import { PrettyRating } from '../../src/components/pretty-rating';

Enzyme.configure({ adapter: new Adapter() });

describe('<PrettyRating />', () => {
	let rating: number;
	let icons: AllowIconsType;

	beforeEach(() => {
		rating = 3.5;
		icons = {
			complete: faStar,
			half: faStarHalfAlt,
			empty: 'star',
		};
	});

	it('Should create Snapshot', () => {
		const component = renderer.create(
			<PrettyRating rating={rating} icons={icons} />,
		);

		const snapshot = component.toJSON();
		expect(snapshot).toMatchSnapshot();
	});
});
