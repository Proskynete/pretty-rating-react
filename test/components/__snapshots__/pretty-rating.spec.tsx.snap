// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PrettyRating /> Should create Snapshot 1`] = `
<div>
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-star fa-w-18 "
    data-icon="star"
    data-prefix="fas"
    focusable="false"
    role="img"
    style={
      Object {
        "color": "#000",
      }
    }
    viewBox="0 0 576 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-star fa-w-18 "
    data-icon="star"
    data-prefix="fas"
    focusable="false"
    role="img"
    style={
      Object {
        "color": "#000",
      }
    }
    viewBox="0 0 576 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-star fa-w-18 "
    data-icon="star"
    data-prefix="fas"
    focusable="false"
    role="img"
    style={
      Object {
        "color": "#000",
      }
    }
    viewBox="0 0 576 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-star-half-alt fa-w-17 "
    data-icon="star-half-alt"
    data-prefix="fas"
    focusable="false"
    role="img"
    style={
      Object {
        "color": "#000",
      }
    }
    viewBox="0 0 536 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M508.55 171.51L362.18 150.2 296.77 17.81C290.89 5.98 279.42 0 267.95 0c-11.4 0-22.79 5.9-28.69 17.81l-65.43 132.38-146.38 21.29c-26.25 3.8-36.77 36.09-17.74 54.59l105.89 103-25.06 145.48C86.98 495.33 103.57 512 122.15 512c4.93 0 10-1.17 14.87-3.75l130.95-68.68 130.94 68.7c4.86 2.55 9.92 3.71 14.83 3.71 18.6 0 35.22-16.61 31.66-37.4l-25.03-145.49 105.91-102.98c19.04-18.5 8.52-50.8-17.73-54.6zm-121.74 123.2l-18.12 17.62 4.28 24.88 19.52 113.45-102.13-53.59-22.38-11.74.03-317.19 51.03 103.29 11.18 22.63 25.01 3.64 114.23 16.63-82.65 80.38z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-star fa-w-18 "
    data-icon="star"
    data-prefix="fas"
    focusable="false"
    role="img"
    style={
      Object {
        "color": "#000",
      }
    }
    viewBox="0 0 576 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M259.3 17.8L194 150.2 47.9 171.5c-26.2 3.8-36.7 36.1-17.7 54.6l105.7 103-25 145.5c-4.5 26.3 23.2 46 46.4 33.7L288 439.6l130.7 68.7c23.2 12.2 50.9-7.4 46.4-33.7l-25-145.5 105.7-103c19-18.5 8.5-50.8-17.7-54.6L382 150.2 316.7 17.8c-11.7-23.6-45.6-23.9-57.4 0z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
</div>
`;
